### PR TITLE
Longhorn v1.1.1 release for branch "dev-v2.5-source"

### PR DESCRIPTION
**Pull request by cclhsu**
Longhorn v1.1.1 dev 2.5

Signed-off-by: Clark Hsu <clark.hsu@suse.com>

### CODE REVIEW

**Review by shuo-wu — CHANGES_REQUESTED** [review]
1. Remove the modification for other packages ` packages/rancher-grafana`
2. Should not remove  `packages/longhorn/generated-changes/patch/Chart.yaml.patch` and ` packages/longhorn/generated-changes/patch/README.yaml.patch`. Instead, you need to update them.
3. New CRDs are missing in `packages/longhorn/generated-changes/overlay/crds/crds.yaml`
4. Please rename the commit title. We are updating for dev-v2.5 rather than v2.6. And this is the v1.1.1 release.


Here are the patches I just generate. You can check the difference:
https://github.com/rancher/charts/compare/dev-v2.5-source...shuo-wu:dev-v2.5-source

**Comment by shuo-wu** [comment]
Notice that the image repos (`rancher/longhornio-xxxxxxx`) in `question.yml` are different from those (`rancher/mirrored-longhornio-xxxxxxx`) in `value.yml`. @cbron Should we make them consistent?

File: packages/longhorn/package.yaml
Comment (jenting) [inline packages/longhorn/package.yaml:3]: Use `commit: v1.1.1` would be easier for future maintenance.

File: packages/longhorn/package.yaml
Comment (shuo-wu) [inline packages/longhorn/package.yaml:5]: This version should be 00.

File: packages/longhorn/generated-changes/patch/questions.yml.patch
Comment (shuo-wu) [inline packages/longhorn/generated-changes/patch/questions.yml.patch:29]: Why removing this?

**Review by shuo-wu — CHANGES_REQUESTED** [review]

File: packages/longhorn/generated-changes/patch/questions.yml.patch
Comment (cclhsu) [reply packages/longhorn/generated-changes/patch/questions.yml.patch:29]: We need to remove it?

**Review by nickgerace — APPROVED** [review]
This looks fine to me from a Helm syntax perspective. The CRDs are supposed to be in the overlay though? They do not come with the chart?

**Comment by cbron** [comment]
@shuo-wu / @yasker is this aiming to get into 2.5.8 ? We are already in the late QA stages for that.

**Comment by yasker** [comment]
@cbron Yes. Longhorn v1.1.1 has been released. The chart should be ready very soon. cc @innobead

File: packages/longhorn/package.yaml
Comment (cclhsu) [reply packages/longhorn/package.yaml:3]: Though unusually, tag can be re-tag to a different commit hash and content.  This might cause some confusion.

Comment (shuo-wu) [reply packages/longhorn/package.yaml:3]: Ideally, the tag is immutable from the perspective of project management. You don't need to worry about that part.

**Review by jenting — APPROVED** [review]
LGTM

**Review by shuo-wu — DISMISSED** [review]
LGTM

**Comment by shuo-wu** [comment]
> This looks fine to me from a Helm syntax perspective. The CRDs are supposed to be in the overlay though? They do not come with the chart?

Yes. The CRDs are in the overlay right now. Please take a final check.

File: packages/longhorn/package.yaml
Comment (cbron) [reply packages/longhorn/package.yaml:3]: Our tooling doesn't support tag yet, just sha.

File: packages/longhorn/generated-changes/patch/values.yaml.patch
Comment (cbron) [inline packages/longhorn/generated-changes/patch/values.yaml.patch:29]: Can't merge until this image exists

**Comment by cbron** [comment]
Issue: https://github.com/rancher/rancher/issues/32369

File: packages/longhorn/generated-changes/patch/questions.yml.patch
Comment (cbron) [inline packages/longhorn/generated-changes/patch/questions.yml.patch:53]: I think all the images in this file should default to the `mirrored-` ones. I'm curious if this default has been causing people to have problems with airgap. Whatever is in values.yaml is what gets added to airgap images.txt.

File: packages/longhorn/generated-changes/overlay/crds/crds.yaml
Comment (cbron) [inline packages/longhorn/generated-changes/overlay/crds/crds.yaml:3]: I didn't deep dive into the CRD's or the version to ensure they were correct, I am assuming QA will find those things.

**Comment by innobead** [comment]
cc @khushboo-rancher

**Comment by meldafrawi** [comment]
LGTM

File: packages/longhorn/generated-changes/patch/Chart.yaml.patch
Comment (PhanLe1010) [inline packages/longhorn/generated-changes/patch/Chart.yaml.patch:7]: We want to support Longhorn installation in Rancher windows cluster. So we may want to remove this annotation: `catalog.cattle.io/os: linux`

Comment (cbron) [reply packages/longhorn/generated-changes/patch/Chart.yaml.patch:7]: Yes if it supports both windows and linux you would remove that annotation

File: packages/longhorn/generated-changes/patch/values.yaml.patch
Comment (cbron) [reply packages/longhorn/generated-changes/patch/values.yaml.patch:29]: exists now

File: packages/longhorn/generated-changes/patch/Chart.yaml.patch
Comment (innobead) [reply packages/longhorn/generated-changes/patch/Chart.yaml.patch:7]: Let's wait for @cclhsu update before merging.

Comment (yasker) [reply packages/longhorn/generated-changes/patch/Chart.yaml.patch:7]: Just to be clear. Longhorn doesn't support Windows. The support we're adding is to allow Longhorn to be installed in the cluster with mixed Linux and WIndows node.

So we should NOT remove this annotation. @cbron @PhanLe1010 please confirm the understanding is right here.

Comment (cbron) [reply packages/longhorn/generated-changes/patch/Chart.yaml.patch:7]: Windows here would imply that it works in some way with the windows nodes. Both logging and monitoring interact with the windows nodes. If longhorn is not interacting with the windows nodes and just tolerates them then keep the Linux annotation. 

Also wait to merge until I or another committer approve, there are a lot of requirements around charts in rancher releases and I need to ensure they are met.

File: packages/longhorn/package.yaml
Comment (jenting) [reply packages/longhorn/package.yaml:3]: In Git, the tag implies the SHA commit ID.
I tried to use the tag, it works the same as using SHA commit ID.

File: packages/longhorn/generated-changes/patch/Chart.yaml.patch
Comment (PhanLe1010) [reply packages/longhorn/generated-changes/patch/Chart.yaml.patch:7]: @yasker @cbron 

Yes, Longhorn doesn't run on Windows nodes. We want to allow the user to install Longhorn on Linux nodes in a cluster that have both Linux and Windows nodes.

If we don't the remove `catalog.cattle.io/os: linux` annotation, Longhorn chart will not be shown in the new Rancher UI when the cluster have both Windows nodes and Linux nodes. So, user cannot install Longhorn even if their cluster have some Linux nodes.

Before removing the `catalog.cattle.io/os: linux` annotation, new Rancher UI shows:
Screen Shot 2021-04-27 at 8 25 05 PM

After removing the `catalog.cattle.io/os: linux` annotation, new Rancher UI shows:
Screen Shot 2021-04-27 at 8 24 45 PM

Comment (innobead) [reply packages/longhorn/generated-changes/patch/Chart.yaml.patch:7]: > My understanding is that we still show Longhorn chart in Windows cluster. When users install Longhorn, they need to specify toleration and node selector so that Longhorn components is only deployed on the Linux nodes as described in the documentation at https://longhorn.io/docs/1.1.1/advanced-resources/deploy/rancher_windows_cluster/

This is what I expect. @yasker any further concerns?

Comment (innobead) [reply packages/longhorn/generated-changes/patch/Chart.yaml.patch:7]: BTW, this frontend implementation is from the below snippets, so obviously if you don't remove this annotation, there is no chance to install longhorn from the hybrid (windows+linux) cluster. (note: the OS annotation is exclusive or empty to allow in the hybrid cluster mode)

https://github.com/rancher/dashboard/blob/176ffcedfd57af42dab27b5c9c7723e8eab1590f/store/catalog.js#L527-L533 and https://github.com/rancher/dashboard/blob/176ffcedfd57af42dab27b5c9c7723e8eab1590f/store/catalog.js#L477-L491

Comment (cclhsu) [reply packages/longhorn/generated-changes/patch/Chart.yaml.patch:7]: @khushboo-rancher @meldafrawi I have updated the test branch  to remove the `linux` annotation.
```
repo: https://github.com/cclhsu/charts
branch: dev-v2.5
```
Please help validate.  Thanks!

Comment (yasker) [reply packages/longhorn/generated-changes/patch/Chart.yaml.patch:7]: @PhanLe1010 I see. It makes sense. Do we have documents on how users can deploy Longhorn in a mixed environment?

@cbron Can you confirm this is expected behavior? So adding the label means it's going to be installed in Linux only cluster. And if we want to support an environment mixed with Windows and Linux, then we need to remove the annotation and ask users to add node selectors to the chart.

Comment (khushboo-rancher) [reply packages/longhorn/generated-changes/patch/Chart.yaml.patch:7]: @cbron Is there a middle way to have option to install Longhorn on a Linux or mixed cluster and doesn't show this for a cluster having only windows nodes?

Comment (PhanLe1010) [reply packages/longhorn/generated-changes/patch/Chart.yaml.patch:7]: >  Do we have documents on how users can deploy Longhorn in a mixed environment?
Yeah, we do have the document at https://longhorn.io/docs/1.1.1/advanced-resources/deploy/rancher_windows_cluster/

Comment (vincent99) [reply packages/longhorn/generated-changes/patch/Chart.yaml.patch:7]: @khushboo-rancher There is no such thing as a cluster with only Windows nodes; you need Linux for the system services.  That's why the filtering is just by "is the cluster Windows or not".

Comment (cbron) [reply packages/longhorn/generated-changes/patch/Chart.yaml.patch:7]: This is technically ok, as long as users don't get confused. They might see it show up in a windows cluster and expect it to work on windows nodes. There will be no way for them to know which way it works.

Comment (yasker) [reply packages/longhorn/generated-changes/patch/Chart.yaml.patch:7]: TLDR: After talking with @cbron , we agree we should keep the annotation there to prevent Longhorn from showing up for the Windows/Linux mixed cluster.

There were some misunderstandings regarding the original request filed by @cbron at https://github.com/rancher/rancher/issues/29220 . The requirement is: Any Rancher featured chart shouldn't be immediately broken when installing in a Windows/Linux mixed cluster, e.g. trying to run a Linux container on top of Windows nodes.

However, Rancher also doesn't recommend running any featured charts for a system it's not designed for. In this case, obviously, Longhorn isn't designed for Windows, so we shouldn't remove the annotation.

But if Rancher doesn't recommend that, why is it required? It's because users can still deploy Helm chart without Rancher UI, e.g. via API/Terraform/Helm. Those ways are not blocked in any sense. There are users who want to do that. But since it's not a common case to deploy Linux specific workload in a Windows/Linux mixed cluster, as the current policy, we don't encourage the user to do it and we want to make sure user knows what they're doing. If they choose to deploy the chart via API/Terraform/Helm, we reasonably believe that they know what they're doing.

If a chart just shows up in user's Windows/Linux mixed cluster, the users might just click and deploy it, then error out - since they need to set related `NodeSelector` and `TaintToleration` manually. It's not a good experience anyway. So it's easier to just not showing it unless the users explicitly deploy it via other means.

Longhorn is not designed for Windows. So we think it's better if we keep the annotation `catalog.cattle.io/os: linux`.

@PhanLe1010 @cclhsu Let me know what do you think.

Comment (PhanLe1010) [reply packages/longhorn/generated-changes/patch/Chart.yaml.patch:7]: Thanks @yasker ! I agree with the conclusion

**Review by cbron — COMMENTED** [review]
Lgtm. After annotation is back in we can merge.

File: packages/longhorn/package.yaml
Comment (cbron) [inline packages/longhorn/package.yaml:5]: This should be 01

**Review by cbron — CHANGES_REQUESTED** [review]
Actually 1 more minor thing then good.

File: packages/longhorn/package.yaml
Comment (yasker) [reply packages/longhorn/package.yaml:5]: Just to clarify @cbron , what does this `releaseCandidateVersion` mean? Also, does it apply to `packageVersion` too?

Comment (cbron) [reply packages/longhorn/package.yaml:5]: Package version of 00 will turn 1.1.1.1 into 1.1.1.100 which is correct. 
releaseCandidateVersion of 01 will turn 1.1.1.100 into 1.1.1.100-rc01, which is also correct

Comment (yasker) [reply packages/longhorn/package.yaml:5]: This is the formal release, so it shouldn't be `rc01`. so the original PR looks good?

Comment (cclhsu) [reply packages/longhorn/package.yaml]: changed.

**Comment by cclhsu** [comment]
I have added back linux annotation, changed releaseCandidateVersion, and squashed the commits.

File: packages/longhorn/package.yaml
Comment (cbron) [reply packages/longhorn/package.yaml]: For right now everything in feature charts (this branch) has an rc until it goes live. That transition happens automatically on rancher release. That process will be going away in the future.

**Review by cbron — APPROVED** [review]

**Review by cbron — APPROVED** [review]

**Pull request merged**
